op-e2e/system_fpp_test.go

```diff
@@ -301,18 +306,29 @@ func testFaultProofProgramScenario(t *testing.T, ctx context.Context, sys *Syste
 	// Should be able to rerun in offline mode using the pre-fetched images
 	fppConfig.L1URL = ""
 	fppConfig.L2URL = ""
-	err = opp.FaultProofProgram(ctx, log, fppConfig)
+	wait = make(chan error, 1)
+	stop, err = opp.FaultProofProgram(log, fppConfig, func(err error) {
+		wait <- err
+	})
 	require.NoError(t, err)
+	require.NoError(t, <-wait)
+	require.NoError(t, stop.Stop(context.Background()))
 
 	// Check that a fault is detected if we provide an incorrect claim
 	t.Log("Running fault proof with invalid claim")
 	fppConfig.L2Claim = common.Hash{0xaa}
-	err = opp.FaultProofProgram(ctx, log, fppConfig)
+	wait = make(chan error, 1)
+	stop, err = opp.FaultProofProgram(log, fppConfig, func(err error) {
+		wait <- err
+	})
+	require.NoError(t, err)
+	result := <-wait
 	if s.Detached {
-		require.Error(t, err, "exit status 1")
+		require.Error(t, result, "exit status 1")
 	} else {
-		require.ErrorIs(t, err, driver.ErrClaimNotValid)
+		require.ErrorIs(t, result, driver.ErrClaimNotValid)
 	}
+	require.ErrorIs(t, stop.Stop(context.Background()), driver.ErrClaimNotValid, "error is sticky on close")
```

Comment:
This new `ErrorIs` check is what's causing the detached tests to fail.

There's two possibilities here:

1. It seems perfectly reasonable to me that `Stop()` would return no error if the process had already completed. I'm not convinced it _should_ return the error that was already passed to `onComplete` given there wasn't any issue stopping. We do need to keep the behaviour consistent between detached and in-process modes though.
2. If we really do want it to return an error, we shouldn't be expecting to get `ErrClaimNotValid` when using detached mode - the specific error type is lost.  In detached mode the test just expects an error. We should actually check something about the error message text there but currently it doesn't because it uses `Error` instead of `ErrorContains`.  So we could just make that section:
```
	if s.Detached {
		require.ErrorContains(t, result, "exit status 1")
		require.ErrorContains(t, stop.Stop(context.Background()), "already exited with non-zero exit code: 1")
	} else {
		require.ErrorIs(t, result, driver.ErrClaimNotValid)
		require.ErrorIs(t, stop.Stop(context.Background()), driver.ErrClaimNotValid, "error is sticky on close")
	}
```

This isn't exactly the same because the error passed to `onComplete` would be a `ExitError` and then its a custom error returned from `Stop` but it is consistent about returning an error.

Reply:
Going with option 2 to keep things simple and get this merged.

Reply:
Fixed this; cleaned up the error-awaiting, and updated the sub-process and in-process client-execution functions to not repeat the execution-result error when stopping the program-server.